src/simulation/Benchmark.cpp

```diff
@@ -0,0 +1,431 @@
+}
+
+bool
+Benchmark::generateLoadForBenchmark(Application& app, uint32_t txRate,
+                                    Metrics& metrics)
+{
+    CLOG(TRACE, LOGGER_ID) << "Generating " << txRate
+                           << " transaction(s) per step";
+
+    mBenchmarkTimeContext->Stop();
+
+    for (uint32_t it = 0; it < txRate; ++it)
+    {
+        uint32_t ledgerNum = app.getLedgerManager().getLedgerNum();
+        auto tx = mSampler->createTransaction();
+
+        mBenchmarkTimeContext->Reset();
+
+        if (!tx->execute(app))
+        {
+            CLOG(ERROR, LOGGER_ID) << "Error while executing a transaction: "
+                                      "transaction was rejected";
+            return false;
+        }
+        mBenchmarkTimeContext->Stop();
+        metrics.txsCount.inc();
+    }
+
+    mBenchmarkTimeContext->Reset();
+
+    CLOG(TRACE, LOGGER_ID) << txRate
+                           << " transaction(s) generated in a single step";
+
+    return true;
+}
+
+void
+Benchmark::scheduleLoad(Application& app, std::function<bool()> loadGenerator,
+                        std::chrono::milliseconds stepTime)
+{
+    VirtualTimer& timer = getTimer(app.getClock());
+    timer.expires_from_now(stepTime);
+    timer.async_wait(
+        [this, &app, loadGenerator, stepTime](asio::error_code const& error) {
+            if (error)
+            {
+                return;
+            }
+            if (loadGenerator())
+            {
+                this->scheduleLoad(app, loadGenerator, stepTime);
+            }
+        });
+}
+
+VirtualTimer&
+Benchmark::getTimer(VirtualClock& clock)
+{
+    if (!mLoadTimer)
+    {
+        mLoadTimer = make_unique<VirtualTimer>(clock);
+    }
+    return *mLoadTimer;
+}
+
+Benchmark::BenchmarkBuilder::BenchmarkBuilder(Hash const& networkID)
+    : mInitialize(false)
+    , mPopulate(false)
+    , mTxRate(0)
+    , mAccounts(0)
+    , mNetworkID(networkID)
+{
+}
+
+Benchmark::BenchmarkBuilder&
+Benchmark::BenchmarkBuilder::setNumberOfInitialAccounts(uint32_t accounts)
+{
+    mAccounts = accounts;
+    return *this;
+}
+
+Benchmark::BenchmarkBuilder&
+Benchmark::BenchmarkBuilder::setTxRate(uint32_t txRate)
+{
+    mTxRate = txRate;
+    return *this;
+}
+
+Benchmark::BenchmarkBuilder&
+Benchmark::BenchmarkBuilder::initializeBenchmark()
+{
+    mInitialize = true;
+    return *this;
+}
+
+Benchmark::BenchmarkBuilder&
+Benchmark::BenchmarkBuilder::populateBenchmarkData()
+{
+    mPopulate = true;
+    return *this;
+}
+
+std::unique_ptr<Benchmark>
+Benchmark::BenchmarkBuilder::createBenchmark(Application& app) const
+{
+    auto sampler = make_unique<ShuffleLoadGenerator>(mNetworkID);
+    if (mPopulate)
+    {
+        prepareBenchmark(app, *sampler);
+    }
+    if (mInitialize)
+    {
+        sampler->initialize(app, mAccounts);
+    }
+    setMaxTxSize(app.getLedgerManager(), MAXIMAL_NUMBER_OF_TXS_PER_LEDGER);
+    app.getHerder().triggerNextLedger(app.getLedgerManager().getLedgerNum());
```

Comment:
without this line core was not submitting anything after accounts initialization, I have to restart it again with forcescp.

Reply:
it goes back to what you expect people to do. I suspect that this is because you started to write benchmark as a self-contained method that would do everything; but as you broke it down into stages this doesn't hold anymore.
For example, with the new approach, you should not have to repopulate accounts etc between benchmark runs for example

Reply:
At least triggerNextLedger call should be inside setMaxTxSize.
In normal run triggerNextLedger is called by ledgerClosed that is called by valueExternalized from HerderImpl.

What about just setting configuration value Config::DESIRED_MAX_TX_PER_LEDGER? That should do what you need.